### PR TITLE
[Playground] Examples CI restore

### PR DESCRIPTION
**Pull request by eantyshev**
It was disabled by #24121

For Python SDK we introduce the building of a latest SDK base image `apache/beam_python3.7_sdk`

addresses #24120

1. introduce caching for Gradle
2. Go, Java dockerfiles: bump BEAM_VERSION to the latest released 2.42.0
3. `-Psdk-tag` build arg for Playground Python runner
4. Small fix to ToB integration trigger
5. Make ToB/Playground Examples CI run on `pull_request`, not on `push`

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by olehborysevych** [comment]
LGMT

File: .github/workflows/playground_examples_ci.yml [hunk 1]
```diff
@@ -16,7 +16,15 @@
 name: Playground Examples CI
 
 on:
-  workflow_dispatch:
```
Comment (damccorm) [inline .github/workflows/playground_examples_ci.yml:19]: Any reason not to keep workflow_dispatch as a trigger as well?

File: .github/workflows/playground_examples_ci_reusable.yml [hunk 2]
```diff
@@ -116,18 +117,32 @@ jobs:
...
+      # TODO build from the current ./sdks/... for Java and Go
       - name: Build SDK Backend Docker image
         run: |
              set -uex
-             opts="-Pdocker-tag=$DOCKERTAG"
-             if [ "$SDK" == "java" ]; then
-              opts="$opts -Pbase-image=apache/beam_java8_sdk:$BEAM_VERSION"
+             opts=" -Pdocker-tag=$DOCKERTAG"
+             if [ "$SDK" == "python" ]; then
+                opts="$opts -Psdk-tag=$SDK_TAG"
```
Comment (damccorm) [inline .github/workflows/playground_examples_ci_reusable.yml:139]: This is probably cleaner/more extensible as `if [ "$SDK_TAG" != "" ]; then` - it took me a bit to piece together why this was gated on python and requires knowledge of the previous step

**Review by damccorm — COMMENTED** [review]
Mostly LGTM, just a few small comments

**Comment by eantyshev** [comment]
check for undefined env variables needs to be disabled for that step

**Pull request merged**